app/models/portfolio.rb

```diff
@@ -22,7 +22,9 @@ def add_portfolio_item(portfolio_item)
   end
 
   def metadata
-    {:user_capabilities => user_capabilities,
-     :shared            => self.access_control_entries.any?}
+    {:user_capabilities        => user_capabilities,
+     :groups_shared_count      => self.access_control_entries.collect(&:group_uuid).uniq.count,
+     :approval_processes_count => self.tags.select { |t| t.name == "workflows" && t.namespace == "approval"}.count,
+     :product_count            => self.portfolio_items.count}
```

Comment:
While I'm not against the concept I would like to see if we can get these to be pure DB queries which only return the count instead of requiring the data to be returned and using `collect` and `select` just to get the count.  The logic should probably be moved to private methods as well.

Anything added to `user_capabilities` needs to make performance a priority or it will eventually become and issue.

Personally I think we need to seriously evaluate if each one of these fields is really needed and not just adding properties because we can.  They will be difficult to remove once added.

For example, is it enough to know if the Portfolio is shared and has approval configured without having to determine the count?  Does the count alone provide any real information?

Reply:
I have to strongly agree with @gmcculloug here. These queries should be VERY carefully vetted once agreed upon that they are needed in the first place.  

I would also like the db to pull static information vs the live queries we see here.  That would require a background worker to recompute at a set schedule ( including rules for invalidation. ) This looks like a major performance impediment in the current form.

Reply:
Chatting briefly with @syncrou I get what he is saying here and wanted to add my own thoughts.

Basically the idea is the values for "groups_shared_count" and "approval_processes_count" are static and could be determined and stored, say in a JSON column, when the instance is updated.  Then the read operation just pulls along the column data like any other column.  They are not required to be part of the `metadata` section, but if we think it is cleaner to keep them grouped here we could keep the new json column internal and including the json data in the `metadata` properties.

The `product_count` would likely need to remind dynamic as that can change without the Portfolio object being updated.

@mkanoor What do you think?

Reply:
Here's a quick take on what I was thinking about.  This definitely turns the logic around that we save data once when things get updated which saves a lot of time on the reads.

The first part is adding a `jsonb` column (`statistics`) to the Portfolios table and then some updates to `portfolio.rb` to write stats either when the `update_statistics` method is call or before saving the object.  The `update_statistics` is for outside callers, think service classes, to trigger the update:
```ruby
before_save { self.statistics = stats }

def update_statistics
  update_attributes(:statistics => stats)
end

private

def stats
  {
    :shared_groups      => access_control_entries.select(:group_uuid).distinct.count,
    :approval_processes => tags.where(:name => "workflows", :namespace => "approval").count,
  }
end
```

Then we have two options for how the data gets returned:
1. The API returns `statistics` like any other column of data which makes the `metadata`  method pretty simple, but means we might have to extend the `statistics` data to include the `portfolio_items` property:
```ruby
def metadata
  {:user_capabilities => user_capabilities}
end

def statistics
  self[:statistics].merge(:portfolio_items => portfolio_items.count)
end
```

2.  The API does not return the `statistics` directly, but moves it into the `metadata` column.  Something like this:
```ruby
def metadata
  {:user_capabilities => user_capabilities}.tap do |data|
    data[:statistics] = statistics.merge(:portfolio_items => portfolio_items.count)
  end
end
```

At some point `portfolio_items.count` moves into the static `statistics` column as well.  It all comes down to ensuring the Portfolio object gets updated when some of the related objects get modified.  I think it should be relatively easy to cover those cases if we do with a few active_record callbacks.

Overall, I think it is sound logic to say we are going to read this data a lot more than we are going to write/update it so it makes sense to optimize the data calculations with that in mind. 

The logic I presented here might not all be correct, but I think it is enough to have a good discussion around.